sys/rtt_stdio/rtt_stdio.c

```diff
@@ -0,0 +1,287 @@
+ */
+static mutex_t _rx_mutex = MUTEX_INIT;
+
+/**
+ * @brief buffer holding stdout
+ */
+static char up_buffer   [STDIO_TX_BUFSIZE];
+
+/**
+ * @brief buffer holding stdin
+ */
+static char down_buffer [STDIO_RX_BUFSIZE];
+
+/**
+ * @brief flag that enables stdin polling
+ */
+static char stdin_enabled = 0;
+
+/**
+ * @brief flag that enables stdout blocking/polling
+ */
+static char blocking_stdout = 0;
+
+/**
+ * @brief SEGGER's ring buffer implementation
+ */
+typedef struct {
+    const char* channel_name; /* Optional name. Standard names so far are:
+                                 "Terminal", "VCom" */
+    char*  buf_ptr;           /* Pointer to start of buffer */
+    int32_t buf_size;         /* Buffer size in bytes. Note that one byte is
+                                 lost, as this implementation does not fill up
+                                 the buffer in order to avoid the problem of
+                                 being unable to distinguish between full and
+                                 empty. */
+    volatile int32_t wr_off;  /* Position of next item to be written by either
+                                 host (down-buffer) or target (up-buffer). Must
+                                 be volatile since it may be modified by host
+                                 (down-buffer) */
+    volatile int32_t rd_off;  /* Position of next item to be read by target
+                                 (down-buffer) or host (up-buffer). Must be
+                                 volatile since it may be modified by host
+                                 (up-buffer) */
+    int32_t    flags;         /* Contains configuration flags */
+} rtt_ringbuf_t;
+
+/**
+ * @brief RTT control block which describes the number of buffers available
+ *        as well as the configuration for each buffer. The struct definition
+ *        is fixed, as it is expected by SEGGER's software
+ */
+typedef struct {
+    char        sentinel[16];      /* Initialized to "SEGGER RTT" */
+    int32_t     max_up_buffers;    /* Initialized to 1 */
+    int32_t     max_down_buffers;  /* Initialized to 1 */
+    rtt_ringbuf_t up[1];           /* Up buffers, transferring information up
+                                      from target via debug probe to host */
+    rtt_ringbuf_t down[1];         /* Down buffers, transferring information
+                                      down from host via debug probe to target */
+} segger_rtt_cb_t;
+
+
+/**
+ * @brief The SEGGER Real-Time-Terminal control block (CB)
+ */
+static segger_rtt_cb_t rtt_cb = {
+    "SEGGER RTT",
+    1,
+    1,
+    {{ "Terminal", &up_buffer[0],   sizeof(up_buffer),   0, 0, 0 }},
+    {{ "Terminal", &down_buffer[0], sizeof(down_buffer), 0, 0, 0 }},
+};
+
+
+/**
+ * @brief read bytes from the down buffer. This function does not block.
+ *        The logic here is unmodified from SEGGER's reference, it is just
+ *        refactored to match code style. The string is not null terminated.
```

Comment:
Since we're importing logic unchanged here, was there any license with the SEGGER code?